src/SelfGravity/CPU_Poisson/CPU_ComputeGREP.cpp

```diff
@@ -151,37 +172,40 @@ void CPU_ComputeGREP( const int lv, const double Time, const Profile_t *DensAve,
    } // if ( ! IsConverged )
 
 
-// 3. compute the effective GR potential at the bin center
+// (3) compute the effective GR potential, defined at the left edge of each bin
    EffPot->NBin        = DensAve->NBin;
    EffPot->LogBin      = DensAve->LogBin;
    EffPot->LogBinRatio = DensAve->LogBinRatio;
    EffPot->MaxRadius   = DensAve->MaxRadius;
    EffPot->AllocateMemory();
 
    for (int d=0; d<3;    d++)   EffPot->Center[d] = DensAve->Center[d];
-   for (int b=0; b<NBin; b++)   EffPot->Radius[b] = DensAve->Radius[b];
+   for (int b=0; b<NBin; b++)   EffPot->Radius[b] = Edge_L[b];
 
-// set the outer boundary condition of the potential to -G * M_out / r_out
-   EffPot->Data[NBin-1] = -NEWTON_G * ( Mass_TOV[NBin-1] - Mass_NW[NBin-1] ) / Radius[NBin-1];
+// set the outer boundary condition of the potential to -GM/r at the left edge of last bin
+   EffPot->Data[NBin-1] = -NEWTON_G * ( Mass_TOV[NBin-2] - Mass_NW[NBin-2] ) / EffPot->Radius[NBin-1];
 
    for (int i=NBin-2; i>0; i--)
    {
-      double dr         = Radius[i] - Radius[i+1];
-      double Mass_NW_C  = 0.5 * ( Mass_NW[i] + Mass_NW[i-1] );
-      double Mass_TOV_C = ( 0.5 * ( Mass_TOV[i] + Mass_TOV[i-1] ) + FourPI * CUBE( Radius[i] ) * Pres[i] * _c2 )
-                        * ( 1.0 + ( Engy[i] + Pres[i] ) / ( Dens[i] * c2 ) )
-                        / SQR( 0.5 * ( Gamma_TOV[i] + Gamma_TOV[i-1] ) );
+      const double dr = EffPot->Radius[i] - EffPot->Radius[i+1];
+
+//    compute the integrand using data at the bin center
+//    --> approximate the Mass_NW, Mass_TOV, and Gamma_TOV at the bin center via linear interpolation
+      const double ratio = ( Radius[i] - Edge_L[i] ) / ( Edge_L[i+1] - Edge_L[i] );
+
+      double Mass_NW_C   = Mass_NW  [i-1] + ratio * ( Mass_NW  [i] - Mass_NW  [i-1] );
+      double Mass_TOV_C  = Mass_TOV [i-1] + ratio * ( Mass_TOV [i] - Mass_TOV [i-1] );
+      double Gamma_TOV_C = Gamma_TOV[i-1] + ratio * ( Gamma_TOV[i] - Gamma_TOV[i-1] );
+
+      Mass_TOV_C = ( Mass_TOV_C + FourPI * CUBE( Radius[i] ) * Pres[i] * _c2 )
+                 * ( 1.0 + ( Engy[i] + Pres[i] ) / ( Dens[i] * c2 ) ) / SQR( Gamma_TOV_C );
+
 
       EffPot->Data[i] = EffPot->Data[i+1] - dr * NEWTON_G * ( Mass_NW_C - Mass_TOV_C ) / SQR( Radius[i] );
    }
 
-   double dr         = Radius[0] - Radius[1];
-   double Mass_NW_C  = Mass_NW[0];
-   double Mass_TOV_C = ( Mass_TOV[0] + FourPI * CUBE( Radius[0] ) * Pres[0] * _c2 )
-                     * ( 1.0 + ( Engy[0] + Pres[0] ) / ( Dens[0] * c2 ) )
-                     / SQR( Gamma_TOV[0] );
-
-   EffPot->Data[0] = EffPot->Data[1] - dr * NEWTON_G * ( Mass_NW_C - Mass_TOV_C ) / SQR( Radius[0] ) ;
+// set the potential for the innermost bin
```

Comment:
Clarify the reason of adopting this zero-gradient inner boundary condition.

Reply:
Comments updated.